File: test-errors/src/programmers/helpers/ProtobufUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtobufUtil = void 0;
var ProtobufUtil;
(function (ProtobufUtil) {
    ProtobufUtil.isStaticObject = (obj) => obj.properties.length >= 1 &&
        obj.properties.every((p) => p.key.isSoleLiteral());
    ProtobufUtil.size = (meta) => ProtobufUtil.getAtomics(meta).length +
        meta.arrays.length +
        meta.tuples.length +
        meta.natives.length +
        meta.objects.length +
        meta.maps.length;
    ProtobufUtil.isUnion = (meta) => meta.arrays.length > 1;
    ProtobufUtil.getAtomics = (meta) => {
        const set = new Set();
        if (meta.templates.length)
            set.add("string");
        for (const c of meta.constants)
            if (c.type === "boolean")
                set.add("bool");
            else if (c.type === "bigint")
                set.add("uint64");
            else if (c.type === "number")
                set.add(deduce_numeric_type(c.values));
            else if (c.type === "string")
                set.add("string");
        for (const atomic of meta.atomics)
            if (atomic.type === "boolean")
                set.add("bool");
            else if (atomic.type === "bigint")
                decode_bigint(atomic.tags).forEach((t) => set.add(t));
            else if (atomic.type === "number")
                decode_number(atomic.tags).forEach((t) => set.add(t));
            else if (atomic.type === "string")
                set.add("string");
        return [...set].sort(compare);
    };
    ProtobufUtil.getNumbers = (meta) => {
        const set = new Set();
        for (const c of meta.constants)
            if (c.type === "number")
                set.add(deduce_numeric_type(c.values));
        for (const atomic of meta.atomics)
            if (atomic.type === "number")
                decode_number(atomic.tags).forEach((t) => set.add(t));
        return [...set].sort(compare);
    };
    ProtobufUtil.getBigints = (meta) => {
        const set = new Set();
        for (const c of meta.constants)
            if (c.type === "bigint")
                set.add("uint64");
        for (const atomic of meta.atomics)
            if (atomic.type === "bigint")
                decode_bigint(atomic.tags).forEach((t) => set.add(t));
        return [...set].sort(compare);
    };
    const compare = (x, y) => ATOMIC_ORDER.get(x) - ATOMIC_ORDER.get(y);
})(ProtobufUtil || (exports.ProtobufUtil = ProtobufUtil = {}));
const ATOMIC_ORDER = new Map([
    "bool",
    "int32",
    "uint32",
    "int64",
    "uint64",
    "float",
    "double",
    "string",
].map((str, i) => [str, i]));
const deduce_numeric_type = (values) => values.every((v) => Math.floor(v) === v)
    ? values.every((v) => -2147483648 <= v && v <= 2147483647)
        ? "int32"
        : "int64"
    : "double";
const decode_bigint = (typeTags) => {
    if (typeTags.length === 0)
        return ["int64"];
    const types = new Set();
    for (const row of typeTags) {
        const value = row.find((tag) => tag.kind === "type" &&
            (tag.value === "int64" || tag.value === "uint64"))?.value;
        types.add(value ?? "int64");
    }
    return [...types];
};
const decode_number = (typeTags) => {
    if (typeTags.length === 0)
        return ["double"];
    const types = new Set();
    for (const row of typeTags) {
        const value = row.find((tag) => tag.kind === "type" &&
            (tag.value === "int32" ||
                tag.value === "uint32" ||
                tag.value === "int64" ||
                tag.value === "uint64" ||
                tag.value === "float" ||
                tag.value === "double"))?.value;
        types.add(value ?? "double");
    }
    return [...types];
};

File: packages/typia/native/core/programmers/helpers/ProtobufUtil.go
package helpers

import (
  "fmt"
  "math"
  "strconv"
  "strings"

  nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type protobufUtilNamespace struct{}

var ProtobufUtil = protobufUtilNamespace{}

func (protobufUtilNamespace) IsStaticObject(obj *nativemetadata.MetadataObjectType) bool {
  if len(obj.Properties) == 0 {
    return false
  }
  for _, property := range obj.Properties {
    if property.Key.IsSoleLiteral() == false {
      return false
    }
  }
  return true
}

func (protobufUtilNamespace) Size(meta *nativemetadata.MetadataSchema) int {
  return len(ProtobufUtil.GetAtomics(meta)) +
    len(meta.Arrays) +
    len(meta.Tuples) +
    len(meta.Natives) +
    len(meta.Objects) +
    len(meta.Maps)
}

func (protobufUtilNamespace) GetSequence(tags []nativemetadata.IMetadataTypeTag) *int {
  for _, tag := range tags {
    if tag.Kind != "sequence" {
      continue
    }
    schema, ok := tag.Schema.(map[string]any)
    if ok == false {
      continue
    }
    value, ok := schema["x-protobuf-sequence"]
    if ok == false {
      continue
    }
    if parsed, ok := protobufUtil_toInt(value); ok {
      return &parsed
    }
  }
  return nil
}

func (protobufUtilNamespace) IsUnion(meta *nativemetadata.MetadataSchema) bool {
  return ProtobufUtil.Size(meta) > 1
}

func (protobufUtilNamespace) GetAtomics(meta *nativemetadata.MetadataSchema, union ...map[string]*int) map[string]*int {
  output := protobufUtil_map(union...)
  for _, constant := range meta.Constants {
    if constant.Type == "boolean" {
      output["bool"] = ProtobufUtil.GetSequence(protobufUtil_firstTags(constant.Values))
    } else if constant.Type == "bigint" {
      init := protobufUtil_deduce_bigint_type(protobufUtil_values(constant.Values))
      for _, value := range constant.Values {
        protobufUtil_decode_bigint(output, value.Tags, init)
      }
    } else if constant.Type == "number" {
      init := protobufUtil_deduce_numeric_type(protobufUtil_values(constant.Values))
      for _, value := range constant.Values {
        protobufUtil_decode_number(output, value.Tags, init)
      }
    } else if constant.Type == "string" {
      output["string"] = ProtobufUtil.GetSequence(protobufUtil_firstTags(constant.Values))
    }
  }
  if len(meta.Templates) != 0 {
    output["string"] = ProtobufUtil.GetSequence(protobufUtil_firstTagRow(meta.Templates[0].Tags))
  }
  for _, atomic := range meta.Atomics {
    if atomic.Type == "boolean" {
      output["bool"] = ProtobufUtil.GetSequence(protobufUtil_firstTagRow(atomic.Tags))
    } else if atomic.Type == "bigint" {
      protobufUtil_decode_bigint(output, atomic.Tags, "int64")
    } else if atomic.Type == "number" {
      protobufUtil_decode_number(output, atomic.Tags, "double")
    } else if atomic.Type == "string" {
      output["string"] = ProtobufUtil.GetSequence(protobufUtil_firstTagRow(atomic.Tags))
    }
  }
  return output
}

func (protobufUtilNamespace) GetNumbers(meta *nativemetadata.MetadataSchema, union ...map[string]*int) map[string]*int {
  output := protobufUtil_map(union...)
  for _, constant := range meta.Constants {
    if constant.Type == "number" {
      init := protobufUtil_deduce_numeric_type(protobufUtil_values(constant.Values))
      for _, value := range constant.Values {
        protobufUtil_decode_number(output, value.Tags, init)
      }
    }
  }
  for _, atomic := range meta.Atomics {
    if atomic.Type == "number" {
      protobufUtil_decode_number(output, atomic.Tags, "double")
    }
  }
  return output
}

func (protobufUtilNamespace) GetBigints(meta *nativemetadata.MetadataSchema, union ...map[string]*int) map[string]*int {
  output := protobufUtil_map(union...)
  for _, constant := range meta.Constants {
    if constant.Type == "bigint" {
      init := protobufUtil_deduce_bigint_type(protobufUtil_values(constant.Values))
      for _, value := range constant.Values {
        protobufUtil_decode_bigint(output, value.Tags, init)
      }
    }
  }
  for _, atomic := range meta.Atomics {
    if atomic.Type == "bigint" {
      protobufUtil_decode_bigint(output, atomic.Tags, "int64")
    }
  }
  return output
}

func (protobufUtilNamespace) Compare(x string, y string) int {
  return protobufUtil_atomicOrder[x] - protobufUtil_atomicOrder[y]
}

var protobufUtil_atomicOrder = map[string]int{
  "bool":   0,
  "int32":  1,
  "uint32": 2,
  "int64":  3,
  "uint64": 4,
  "float":  5,
  "double": 6,
  "string": 7,
}

func protobufUtil_map(union ...map[string]*int) map[string]*int {
  if len(union) != 0 && union[0] != nil {
    return union[0]
  }
  return map[string]*int{}
}

func protobufUtil_firstTags(values []*nativemetadata.MetadataConstantValue) []nativemetadata.IMetadataTypeTag {
  if len(values) == 0 || len(values[0].Tags) == 0 {
    return []nativemetadata.IMetadataTypeTag{}
  }
  return values[0].Tags[0]
}

func protobufUtil_firstTagRow(tags [][]nativemetadata.IMetadataTypeTag) []nativemetadata.IMetadataTypeTag {
  if len(tags) == 0 {
    return []nativemetadata.IMetadataTypeTag{}
  }
  return tags[0]
}

func protobufUtil_values(values []*nativemetadata.MetadataConstantValue) []any {
  output := make([]any, 0, len(values))
  for _, value := range values {
    output = append(output, value.Value)
  }
  return output
}

func protobufUtil_deduce_bigint_type(values []any) string {
  for _, value := range values {
    if strings.HasPrefix(fmt.Sprint(value), "-") {
      return "int64"
    }
  }
  return "uint64"
}

func protobufUtil_deduce_numeric_type(values []any) string {
  integers := true
  int32s := true
  for _, value := range values {
    number := protobufUtil_toFloat(value)
    if math.Floor(number) != number {
      integers = false
      break
    }
    if -2147483648 <= number && number <= 2147483647 {
      continue
    }
    int32s = false
  }
  if integers == false {
    return "double"
  }
  if int32s {
    return "int32"
  }
  return "int64"
}

func protobufUtil_decode_bigint(output map[string]*int, tags [][]nativemetadata.IMetadataTypeTag, defaultType string) {
  if len(tags) == 0 {
    output[defaultType] = nil
    return
  }
  for _, row := range tags {
    value := "int64"
    for _, tag := range row {
      if tag.Kind == "type" && (tag.Value == "int64" || tag.Value == "uint64") {
        value = fmt.Sprint(tag.Value)
        break
      }
    }
    output[value] = ProtobufUtil.GetSequence(row)
  }
}

func protobufUtil_decode_number(output map[string]*int, tags [][]nativemetadata.IMetadataTypeTag, defaultType string) {
  if len(tags) == 0 {
    output[defaultType] = nil
    return
  }
  for _, row := range tags {
    value := "double"
    for _, tag := range row {
      if tag.Kind == "type" {
        str := fmt.Sprint(tag.Value)
        if str == "int32" || str == "uint32" || str == "int64" || str == "uint64" || str == "float" || str == "double" {
          value = str
          break
        }
      }
    }
    output[value] = ProtobufUtil.GetSequence(row)
  }
}

func protobufUtil_toInt(value any) (int, bool) {
  switch v := value.(type) {
  case int:
    return v, true
  case int32:
    return int(v), true
  case int64:
    return int(v), true
  case uint:
    return int(v), true
  case uint32:
    return int(v), true
  case uint64:
    return int(v), true
  case float64:
    return int(v), true
  case float32:
    return int(v), true
  case string:
    parsed, err := strconv.Atoi(v)
    return parsed, err == nil
  default:
    return 0, false
  }
}

func protobufUtil_toFloat(value any) float64 {
  switch v := value.(type) {
  case float64:
    return v
  case float32:
    return float64(v)
  case int:
    return float64(v)
  case int32:
    return float64(v)
  case int64:
    return float64(v)
  case uint:
    return float64(v)
  case uint32:
    return float64(v)
  case uint64:
    return float64(v)
  case string:
    parsed, _ := strconv.ParseFloat(v, 64)
    return parsed
  default:
    parsed, _ := strconv.ParseFloat(fmt.Sprint(value), 64)
    return parsed
  }
}
